prescient/data/providers/gmlc_data_provider.py

```diff
@@ -0,0 +1,220 @@
+        time_period_length_minutes: int
+            The duration of each time step
+        model: EgretModel
+            The model where actuals data will be stored
+
+        Notes
+        -----
+        This will store actuals data in the model's existing data arrays, starting
+        at index 0.  If the model's arrays are not big enough to hold all the
+        requested time steps, only those steps for which there is sufficient storage
+        will be saved.  If arrays are larger than the number of requested time 
+        steps, the remaining array elements will be left unchanged.
+
+        Note that this method has the same signature as populate_with_forecast_data.
+        '''
+        self._populate_with_forecastable_data('REAL_TIME', start_time, num_time_periods, time_period_length_minutes, model)
+
+
+    def _populate_with_forecastable_data(self,
+                                         sim_type:str,
+                                         start_time:datetime,
+                                         num_time_periods: int,
+                                         time_period_length_minutes: int,
+                                         model: EgretModel
+                                        ) -> None:
+        end_time = start_time + timedelta(minutes=num_time_periods*time_period_length_minutes)
+        native_frequency = self._cache.minutes_per_period[sim_type]
+        step_ratio = time_period_length_minutes // native_frequency
+        
+        if step_ratio == 1 and len(model.data['system']['time_keys']) == num_time_periods:
+            self._cache.populate_skeleton_with_data(model.data, sim_type, start_time, end_time)
+        else:
+            copy_from = self._cache.generate_model(sim_type, start_time, end_time)
+            _recurse_copy_at_ratio(copy_from.data, model.data, step_ratio)
+            # Fill in the times
+            time_labels = model.data['system']['time_keys']
+            delta = timedelta(minutes=time_period_length_minutes)
+            for i in range(len(time_labels)):
+                dt = start_time + i*delta
+                time_labels[i] = dt.strftime('%Y-%m-%d %H:%M')
+
+def _recurse_copy_at_ratio(src:dict[str, Any], target:dict[str, Any], ratio:int) -> None:
+    ''' Copy every Nth value from a src dict's time_series values into corresponding arrays in a target dict.
+    '''
+    for key, att in src.items():
+        if isinstance(att, dict):
+            if 'data_type' in att and att['data_type'] == 'time_series':
+                src_vals = att['values']
+                if type(target[key]) is dict and target[key]['data_type'] == 'time_series':
+                    # If there is already a value array at the target, fill it in (to preserve array size)
+                    target_vals = target[key]['values']
+                    for s,t in zip(range(0, len(src_vals), ratio), range(len(target_vals))):
+                        target_vals[t] = src_vals[s]
```

Comment:
I'm moderately confused here. Is the implicit assumption that `len(target_vals) == len(src_vals)/ratio`? Python's `zip` function does not check this for you (see https://www.python.org/dev/peps/pep-0618/).

Reply:
We deliberately want it to stop at the shorter of the two sequences, which is what zip does.  The example that comes to mind is when you want to use estimate data for the first 24 hours, then fill in estimates beyond the 24 hours using a heuristic such as repeating the final value (as is done for wind estimates).  In this case the source has 24 hours but the target has more than 24 hours.